src/vm/collection.rs

```diff
@@ -112,28 +112,43 @@ pub trait Collection<VM: VMBinding> {
     /// * `tls_worker`: The thread pointer for the worker thread performing this call.
     fn post_forwarding(_tls: VMWorkerThread) {}
 
-    /// Return the amount of memory (in pages) the VM allocated off-heap which could be released
-    /// by doing GC.  MMTk core will trigger a GC if the amount of allocated memory in MMTk spaces
-    /// plus this amount exceeds a certain limit.
+    /// Return the amount of memory (in bytes) the VM allocated (but not released yet) outside the
+    /// MMTk heap which can be released by doing GC.
+    ///
+    /// MMTk core will add this amount to the amount of memory allocated or reserved in MMTk
+    /// spaces, and will trigger a GC if the sum exceeds a certain limit.
+    ///
+    /// MMTk does not specify what memory this amount should includes.  However, because this
+    /// amount is used to trigger GC, it is advisable to include memory that can be released by
+    /// executing finalizers or other language-specific cleaning-up routines that are executed when
+    /// an in-heap object is dead.  For example, if a language implementation allocates array
+    /// headers in the MMTk heap, but allocates their underlying buffers that hold the actual
+    /// elements using `malloc`, then those buffers should be included in this amount.  When the GC
+    /// finds such an array dead, its finalizer shall `free` the buffer and reduce this amount.
+    ///
+    /// This function is a hint for the MMTk core to trigger GC, therefore does not have to be
```

Comment:
But the "min heap size" is an empirical value, too. It depends on the plan, the number of CPUs (which determines the number of GC threads) and other specifics of the machine the program is running on.  I think it is OK to say it is a hint, given that the exact set of "objects kept alive by heap objects" is not precisely defined.  Another important point is that the VM should not spend too much time computing the exact value because this function is called very often (by `gc_poll`).  An approximate value should be good enough for MMTk to trigger GC promptly and keep the memory footprint under control.

Reply:
I see what you mean. The value does not need to be precise for the sake of performance. But I woudn't say it is a 'hint'. A 'hint' means something like an suggestion. Like Java's `System.gc()` is a hint, we may do or may not do a GC. Julia's `--heap-size-hint` is a hint, the actual heap size can be larger than the given value. But in our case, the number returned by the method is not a hint, and we do count it into our heap size and we will always count it -- that is the contract of this method, the memory is counted into our heap size.

You have another paragraph talking about performance. If you would like to emphasis further on the tradeoff between performance and being precise, you can expand that section.

Reply:
Sounds good.  I'll remove the "hint" part and merge it with the paragraph about performance.